### PR TITLE
Implements `--start-from-copy` option defined in #392

### PR DESCRIPTION
**Pull request by joaomcteixeira**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and you comply with the following criteria:

- [x] You have stick to Python. Talk with us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [ ] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there's a (state) purpose
- [x] code follows our coding style
- [x] You wrote tests for the new code
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [x] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any *install dependencies* unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

* implements `--start-from-copy` option defined in #392

### CODE REVIEW

**Comment by joaomcteixeira** [comment]
I think everything is implemented. It was tricker than thought at first.

Please test it as much as possible, following the docs in #392, using the `--restart` option, the `haddock3-copy` new CLI, and the `haddock3 --start-from-copy` flag.

To install the new CLI, after you pull the branch you need to perform:

```
python setup.py develop --no-deps
```

For those that do not have my fork in your git folder:

```
git remote add joaot https://github.com/joaomcteixeira/haddock3
git fetch joaot
git checkout restart-from-dir
```

Still need to add some tests and documentation, but everything should be working now.

**Review by rvhonorato — APPROVED** [review]
👍🏽

File: src/haddock/gear/prepare_run.py [hunk 1]
```diff
@@ -106,52 +115,100 @@ def setup_run(workflow_path, restart_from=None):
         Whether to erase the previous run folder and reprare from
         scratch. Defaults to `True`.
 
+    start_from_copy : str or Path
+        The path created with `haddock3-copy` to start the run from.
+        Defaults to None, do not use this option.
```
Comment (douweschulte) [inline src/haddock/gear/prepare_run.py:120]: Docs suggestion: 'Default to None, which does not use this option'.

File: src/haddock/gear/prepare_run.py [hunk 2]
```diff
@@ -106,52 +115,100 @@ def setup_run(workflow_path, restart_from=None):
...
+    if not_none(start_from_copy):
+        with suppress(TypeError):
+            start_from_copy = Path(start_from_copy)
 
-    populate_mol_parameters(modules_params)
+        general_params[RUNDIR] = start_from_copy
 
-    # validations
-    validate_modules_params(modules_params)
+    validate_module_names_are_not_mispelled(modules_params)
```
Comment (douweschulte) [inline src/haddock/gear/prepare_run.py:149]: Unrelated: I still like this misspelling of the word `mispelled` (should be `misspelled`).

File: src/haddock/gear/prepare_run.py [hunk 3]
```diff
@@ -106,52 +115,100 @@ def setup_run(workflow_path, restart_from=None):
...
+        The path created with `haddock3-copy` to start the run from.
+        Defaults to None, do not use this option.
+
     Returns
     -------
     tuple of two dicts
         A dictionary with the parameters for the haddock3 modules.
         A dictionary with the general run parameters.
     """
+    from_scratch = restart_from is None and start_from_copy is None
```
Comment (douweschulte) [inline src/haddock/gear/prepare_run.py:128]: Maybe it is better to move this to just before where it is used (just before line 153 according to Github).

File: src/haddock/gear/start_from_copy.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    Example
+    -------
+    >>> steps = ['0_topoaa', '4_flexref']
+    >>> zero_fill.set_zerofill_number(len(steps))
+    >>> copy_renum_step_folders('run1', 'newrun', steps)
+
+    The initial structure::
+
+        run1/
+            1_rigidbody/
```
Comment (douweschulte) [inline src/haddock/gear/start_from_copy.py:95]: Do I read it wrong if I think that `0_topoaa` is missing here in the initial structure?

File: src/haddock/libs/libfunc.py [hunk 5]
```diff
@@ -24,3 +25,8 @@ def none(*ignore, **everything):
 def nan(*ignore, **everything):
     """Give False regardless of the input."""
     return float('nan')
+
+
+def not_none(value):
+    """Give True if value is None, or False otherwise."""
```
Comment (douweschulte) [inline src/haddock/libs/libfunc.py:31]: It seem that the docstring here is missing a `not`.

**Review by douweschulte — APPROVED** [review]
Nice work! I left some small comments inline. I know you are still working on the docs but still pointed out what I found.

File: src/haddock/gear/prepare_run.py [hunk 6]
```diff
@@ -315,28 +369,57 @@ def create_data_dir(run_dir):
...
-    params['topoaa']['molecules'] = list(map(Path, params['molecules']))
+    topoaa_params['molecules'] = list(map(Path, molecules))
 
 
-def copy_input_files_to_data_dir(data_dir, modules_params):
-    """Copy files to data directory."""
-    new_mp = deepcopy(modules_params)
+def copy_molecules_to_data_dir(data_dir, topoaa_params):
+    """
+    Copy molecules to data directory and topoaa parameters.
```
Comment (mgiulini) [inline src/haddock/gear/prepare_run.py:379]:
```suggestion
    Copy molecules (to data directory) and topoaa parameters.
```

File: tests/test_gear_start_from_copy.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+"""Test start from copy gear."""
```
Comment (mgiulini) [inline tests/test_gear_start_from_copy.py:1]:
```suggestion
"""Test start_from_copy gear."""
```

**Review by mgiulini — APPROVED** [review]

File: src/haddock/gear/prepare_run.py (hunk 2 above)
Comment (joaomcteixeira) [reply src/haddock/gear/prepare_run.py:149]: It has a very clear origin in my brain:
`mis-` is used to assign a "bad", "wrong" behavior to words -> https://www.macmillandictionary.com/dictionary/british/mis

That's why I constantly type `mispelled`, from wrongly-spelled. Clearly, it is wrong. The correct is `misspelled`. What should it be? Missed the correct spelling? :-P Thanks again @douweschulte

**Comment by joaomcteixeira** [comment]
Remember to do `python setup.py develop --no-deps` after `git pull` to install the new client.

**Comment by joaomcteixeira** [comment]
Thanks all for your reviews :rocket:

**Pull request merged**